### PR TITLE
Perf env setup script

### PR DESCRIPTION
**Pull request by ldemailly**
Sets up a cluster and 
In VM and in K8s fortio for perf measurement

Installs istio
```release-note
In k8s performance characterization scripts
```

### CODE REVIEW

File: tools/setup_perf_cluster.sh [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+function run_on_vm() {
+  echo "*** Remote run: \"$1\""
+  execute gcloud compute ssh $VM_NAME $GCP_OPTS --command "$1"
+}
+
+function setup_vm() {
+  run_on_vm 'sudo add-apt-repository ppa:gophers/archive > /dev/null && sudo apt-get update > /dev/null && sudo apt-get install --no-install-recommends -y golang-1.8-go && mv .bashrc .bashrc.orig && (echo "export PATH=/usr/lib/go-1.8/bin:\$PATH:~/go/bin"; cat .bashrc.orig) > ~/.bashrc'
+  execute gcloud compute instances add-tags $VM_NAME $GCP_OPTS --tags port8080
+  execute gcloud compute --project $PROJECT firewall-rules create allow-8080 --direction=INGRESS --action=ALLOW --rules=tcp:8080 --target-tags=port8080
```
Comment (ldemailly) [inline tools/setup_perf_cluster.sh]: GKE keeps deleting that entry somehow...

File: tools/setup_perf_cluster.sh [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  execute gcloud compute instances create $VM_NAME $GCP_OPTS --machine-type $MACHINE_TYPE --image $VM_IMAGE
+}
+
+function run_on_vm() {
+  echo "*** Remote run: \"$1\""
+  execute gcloud compute ssh $VM_NAME $GCP_OPTS --command "$1"
+}
+
+function setup_vm() {
+  run_on_vm 'sudo add-apt-repository ppa:gophers/archive > /dev/null && sudo apt-get update > /dev/null && sudo apt-get install --no-install-recommends -y golang-1.8-go && mv .bashrc .bashrc.orig && (echo "export PATH=/usr/lib/go-1.8/bin:\$PATH:~/go/bin"; cat .bashrc.orig) > ~/.bashrc'
```
Comment (mandarjog) [inline tools/setup_perf_cluster.sh]: we use go 1.9 everywhere. Fortio should use it too.

Comment (ldemailly) [reply tools/setup_perf_cluster.sh]: oh, I didn't know, when did we switch ? have we already made 1.8 incompatible changes ? 
would be good I compare 1.8 and 1.9 before we do switch
(and I had been staying at 1.8.x for fortio to be compatible with the rest of istio)

File: tools/setup_perf_cluster.sh [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+GCP_OPTS="--project $PROJECT --zone $ZONE"
+
+function execute() {
+  echo "### Running:" "$@"
+  "$@"
+}
+
+function create_cluster() {
+  execute gcloud container clusters create $CLUSTER_NAME $GCP_OPTS --machine-type=$MACHINE_TYPE --num-nodes=$NUM_NODES --no-enable-legacy-authorization
```
[the remark this answers is not in the document]
Comment (ldemailly) [reply tools/setup_perf_cluster.sh]: no this is by design, I want the "current default" and as we chatted we'll run this periodically and compare the numbers before we ditch the older setup

File: tools/setup_perf_cluster.sh (hunk 2 above)
Comment (mandarjog) [reply tools/setup_perf_cluster.sh]: If you want to ensure that the source remains 1.8 compatible, you can do that. But still use 1.9 compiler.

File: tools/setup_perf_cluster.sh [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+function get_vm_ip() {
+  VM_IP=$(gcloud compute instances describe $VM_NAME $GCP_OPTS |grep natIP|awk -F": " '{print $2}')
+  echo "+++ VM Ip is $VM_IP - visit http://$VM_IP/fortio/"
+}
+
+# assumes run from istio/ (or release) directory
+function install_istio() {
+  # Use the non debug ingress and remove the -v "2"
+  execute sh -c 'sed -e "s/_debug//g" install/kubernetes/istio-auth.yaml | egrep -v -e "- (-v|\"2\")" | kubectl apply -f -'
```
Comment (ldemailly) [inline tools/setup_perf_cluster.sh]: cc @suryadu

**Comment by ldemailly** [comment]
@yutongz @bochunz - this is I think useful as a start - PTAL / I'd like to merge

@mandarjog with this script you can do the following:
```
source tools/setup_perf_cluster.sh 
export PROJECT=yourproject
# Then
install_istio_svc
install_istio_ingress_rules
install_non_istio_svc
get_non_istio_ingress_ip
get_istio_ingress_ip
...
```
will install fortio svc both kube-injected and not and ingress to access them

File: tools/setup_perf_cluster.sh [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+}
+function run_fortio_test3() {
+  echo "Using istio ingress:"
+  Execute curl "http://$VM_IP/fortio/?json=on&qps=-1&t=30s&c=48&load=Start&url=http://$ISTIO_INGRESS_IP/fortio1/echo"
+}
+
+function setup_vm_all() {
+  update_gcp_opts
+  create_vm
+  #TODO: 'wait for vm to be ready'
```
Comment (yutongz) [inline tools/setup_perf_cluster.sh]: Maybe for now, just do a dummy sleep 5s or so.

File: tools/setup_perf_cluster.sh [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+#
+
+ZONE=${ZONE:-us-east4-b}
+CLUSTER_NAME=${CLUSTER_NAME:-istio-perf}
+MACHINE_TYPE=${MACHINE_TYPE:-n1-highcpu-2}
+NUM_NODES=${NUM_NODES:-6} # SvcA<->SvcB + Ingress + Pilot + Mixer + 1 extra (kube-system)
+VM_NAME=${VM_NAME:-fortio-vm}
+ISTIOCTL=${ISTIOCTL:-istioctl} # to override istioctl from outside of the path
+FORTIO_NAMESPACE=${FORTIO_NAMESPACE:-fortio} # Namespace for non istio app
+ISTIO_NAMESPACE=${ISTIO_NAMESPACE:-istio} # Namespace for istio injected app
```
Comment (yutongz) [inline tools/setup_perf_cluster.sh]:
```
### Running: sh -c istioctl kube-inject --debug=false -n istio -f ./tools/perf_k8svcs.yaml > ./tools/perf_k8svcs_istio.yaml
### Running: kubectl apply -n istio -f ./tools/perf_k8svcs_istio.yaml
Error from server (NotFound): error when creating "./tools/perf_k8svcs_istio.yaml": namespaces "istio" not found
Error from server (NotFound): error when creating "./tools/perf_k8svcs_istio.yaml": namespaces "istio" not found
Error from server (NotFound): error when creating "./tools/perf_k8svcs_istio.yaml": namespaces "istio" not found
Error from server (NotFound): error when creating "./tools/perf_k8svcs_istio.yaml": namespaces "istio" not found
Error from server (NotFound): error when creating "./tools/perf_k8svcs_istio.yaml": namespaces "istio" not found
```

```
➜  istio git:(perf_scripts) ✗ k get ns
NAME           STATUS    AGE
default        Active    5m
fortio         Active    4m
istio-system   Active    4m
kube-public    Active    5m
kube-system    Active    5m
```

Should it be istio-system or something?

Comment (ldemailly) [reply tools/setup_perf_cluster.sh]: good catch, thank you - somehow the create namespace for that one is missing

Comment (ldemailly) [reply tools/setup_perf_cluster.sh]: it's commented out on line 114 so it doesn't fail on rerun

**Pull request merged**